bitwarden_license/src/Scim/Controllers/v2/GroupsController.cs

```diff
@@ -56,59 +56,8 @@ public async Task<IActionResult> Get(Guid organizationId, Guid id)
         [FromQuery] int? count,
         [FromQuery] int? startIndex)
     {
-        string nameFilter = null;
-        string externalIdFilter = null;
-        if (!string.IsNullOrWhiteSpace(filter))
-        {
-            if (filter.StartsWith("displayName eq "))
-            {
-                nameFilter = filter.Substring(15).Trim('"');
-            }
-            else if (filter.StartsWith("externalId eq "))
-            {
-                externalIdFilter = filter.Substring(14).Trim('"');
-            }
-        }
-
-        var groupList = new List<ScimGroupResponseModel>();
-        var groups = await _groupRepository.GetManyByOrganizationIdAsync(organizationId);
-        var totalResults = 0;
-        if (!string.IsNullOrWhiteSpace(nameFilter))
-        {
-            var group = groups.FirstOrDefault(g => g.Name == nameFilter);
-            if (group != null)
-            {
-                groupList.Add(new ScimGroupResponseModel(group));
-            }
-            totalResults = groupList.Count;
-        }
-        else if (!string.IsNullOrWhiteSpace(externalIdFilter))
-        {
-            var group = groups.FirstOrDefault(ou => ou.ExternalId == externalIdFilter);
-            if (group != null)
-            {
-                groupList.Add(new ScimGroupResponseModel(group));
-            }
-            totalResults = groupList.Count;
-        }
-        else if (string.IsNullOrWhiteSpace(filter) && startIndex.HasValue && count.HasValue)
-        {
-            groupList = groups.OrderBy(g => g.Name)
-                .Skip(startIndex.Value - 1)
-                .Take(count.Value)
-                .Select(g => new ScimGroupResponseModel(g))
-                .ToList();
-            totalResults = groups.Count;
-        }
-
-        var result = new ScimListResponseModel<ScimGroupResponseModel>
-        {
-            Resources = groupList,
-            ItemsPerPage = count.GetValueOrDefault(groupList.Count),
-            TotalResults = totalResults,
-            StartIndex = startIndex.GetValueOrDefault(1),
-        };
-        return new ObjectResult(result);
+        var scimListResponseModel = await _getGroupsListQuery.GetGroupsListAsync(organizationId, filter, count, startIndex);
```

Comment:
As with the other commands, let's try to construct the Response in the controller. This looks a bit harder here but I think we can still do it.

1. Return an `IEnumerable<Group>` from the command
2. Then use `Select` to map over those groups and transform them into `ScimGroupResponseModel`s
3. Then construct your `ScimListResponseModel` - noting that all the other properties are either derived from the `IEnumerable<Group>` (TotalResults) or from the request (ItemsPerPage, StartIndex)

This leaves the command to just deal with getting and filtering/paging the groups, which is a nicer separation of responsibilities.

Reply:
I modified it to return a Tuple because `TotalResults` can be bigger than the actual returned result from the query.

Reply:
Good catch, I missed that. I definitely like this approach more, even though returning a tuple can be slightly awkward.